crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -81,14 +82,67 @@ pub(crate) fn infer_scope_types<'db>(db: &'db dyn Db, scope: ScopeId<'db>) -> Sc
     TypeInferenceBuilder::new(db, InferenceRegion::Scope(scope), index, &module).finish_scope()
 }
 
-fn scope_cycle_initial<'db>(_db: &'db dyn Db, scope: ScopeId<'db>) -> ScopeInference<'db> {
-    ScopeInference::cycle_initial(scope)
+fn scope_cycle_recover<'db>(
+    db: &'db dyn Db,
+    id: salsa::Id,
+    previous_inference: &ScopeInference<'db>,
+    inference: &ScopeInference<'db>,
+    _count: u32,
+    _scope: ScopeId<'db>,
+) -> salsa::CycleRecoveryAction<ScopeInference<'db>> {
+    let ScopeInference {
+        expressions,
+        scope,
+        extra,
+    } = inference;
+    let mut expressions = expressions.clone();
```

Comment:
It seems like it would be beneficial if salsa passed the owned current value instead of a reference. I'd have to think about how such an API exactly should look like because we don't want to call `::eq` twice when returning `Iterate` but we also need to ensure that the value hasn't been changed when using `::Iterate`. 

Unless we can find a way to defer cloning in ty unless we hit a divergent value. But i suspect that this is might be harder